scripts/cleanup.js

```diff
@@ -53,124 +67,116 @@ function getImages(region, repo, callback) {
-  });
+  const generic = [];
+  const priority = [];
 
-  let cruftDigests = [];
-  let deployDigests = [];
-
-  for(let img of images) {
-    q.defer(commitType, img.imageTags[0], img.imageDigest);
-  }
+  // Sort oldest to newest, categorize each image
+  images
+    .sort((a, b) => new Date(a.imagePushedAt) - new Date(b.imagePushedAt))
+    .forEach((img) => q.defer(commitType, img.imageTags[0], img.imageDigest, tmpdir));
 
   q.awaitAll((err, data) => {
     if (err) return callback(err);
 
-    for (let d of data) {
-      for (let digest in d) {
-        if (d[digest] === 'commit') cruftDigests.push({ imageDigest: digest });
-        else if (d[digest] !== 'custom') deployDigests.push({ imageDigest: digest });
-      }
-    }
+    // Select generic images and priority images
+    data.forEach((result) => {
+      if (result.type === 'generic')
+        generic.push({ imageDigest: result.digest });
+      if (result.type === 'priority')
+        priority.push({ imageDigest: result.digest });
+    });
 
-  });
+    // We want to leave the repository with MAX_IMAGES number of images in it.
+    let imagesToDelete;
+    let excessImages = images.length - MAX_IMAGES;
+    if (excessImages <= 0) return callback(null, []);
 
-  console.log('cruftDigests: ', cruftDigests.join(','));
-  console.log('deployDigests: ', deployDigests.join(','));
+    // First take old images from generic commits
+    imagesToDelete = []
+      .concat(generic.slice(0, excessImages));
 
-  let digests = [];
-  digests = digests.concat(cruftDigests.slice(0, (digests.length - (MAX_IMAGES - 1))));
+    // Determine whether we still need to delete more images
+    excessImages = excessImages - imagesToDelete.length;
+    if (excessImages <= 0) return callback(null, imagesToDelete);
 
-  if (deployDigests.length > MAX_PRIORITY_IMAGES)
-    digests = digests.concat(deployDigests.slice(0, (deployDigests.length - MAX_PRIORITY_IMAGES)));
+    // Make sure that we always leave at least 50 priority images
+    excessImages = Math.max(
+      Math.min(priority.length - MIN_PRIORITY_IMAGES, excessImages),
+      0
+    );
 
-  console.log('digests ', digests.join(','));
-  return callback(null, digests);
+    // Select the oldest priority images for removal
+    imagesToDelete = imagesToDelete
+      .concat(priority.slice(0, excessImages));
+
+    return callback(null, imagesToDelete);
+  });
 }
 
 module.exports.deleteImages = deleteImages;
-function deleteImages(region, repo, imageIds, callback) {
+function deleteImages(region, repositoryName, images, callback) {
+  const ecr = new AWS.ECR({ region: region });
 
-  let ecr = new AWS.ECR({ region: region });
-  ecr.batchDeleteImage({
-    imageIds: imageIds,
-    repositoryName: repo
-  }, callback);
+  // Must delete images in batches of 100 max
+  const remaining = JSON.parse(JSON.stringify(images));
+  const imageIds = remaining.splice(0, 100);
```

Comment:
Needed to account for the limit in number of images that can be deleted in one request.